**RISK ASSESSMENT FOR AIRCRAFTS IN THE AVIATION SECTOR**

**Problem Statement**

By getting involved in the aviation sector, the company is braodening it's portfolio by buying and managing aircraft for both 
individual and commercial business. The distinct operational and safety issues connected to various aircraft models,however, 
are not well understood. To choose the aircraft with the best safety records and fewest hazards, it's difficult to evaluate 
past safety and performance data. Leadership in the aviation division will be better eqquiped to make wise purchases and put 
risk-reduction plans into action thanks to this analysis.

**Objectives**

**1. Establish a Relationship Between Aircraft Features and Accident Severity**

To examine important aircraft features (Model,Make,Engine Type, Aircraft Damage) in conjuction with the injury metrics
(Total Fatal Injuries, Total Serious Injuries, Total Minor Injuries, Total Uninjured) to identify which models routinely
exhibit lower accident severity rates.

**2. Evaluate the Effects of Weather and Flight on the Severity of Accidents**

Examine the effects of weather(VMC vs IMC) on accident outcomes. To also assertain when the hazards are greatest, 
examine incidents by general flight phases(Take off, Cruise/enoute, Approach/Landing).

**3. Create a Data-Informed Risk Profile to help with Aircraft Investing Decision**

Considering accident fatalities, iinjury severity and aircraft damage to rank aircraft safety. Make recommendations on the 
safest aircraft types for the company's commercial and private aviation investments.

**Data Cleaning and Visualization of The Dataset**

**1. Data Cleaning**

**a) Importing and inspecting data**

We start by import the Pandas library,Numpy library, matplotlib.pyplot library that we'll be using in data cleaning and visualiasation and also load the AviationData using (pd.read_csv).

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [16]:
df = pd.read_csv("AviationData.csv", encoding= "latin-1")
df.replace("-", np.nan, inplace=True)
#used latin-1 beacause utf-18 wasn't able to encode the csv file.
df

C:\Users\User\AppData\Local\Temp\ipykernel_5032\1965686645.py:1: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("AviationData.csv", encoding= "latin-1")


Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude  Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2        Saltville, VA  United States  36.922223 -81.878056          NaN   
3           EUREKA, CA  United States        NaN        NaN          NaN   
4           Canton, OH  United States        NaN        NaN          NaN   
...                ...            ...        ...        ...          ...   
88884    Annapolis, MD  United States        NaN        NaN          NaN   
88885      Hampton, NH  United States        NaN        NaN          NaN   
88886       Payson, AZ  United States    341525N   1112021W          PAN   
88887       Morgan, UT  United States        NaN        NaN          NaN   
88888       Athens, GA  United States        NaN        NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0               N

Calling the first five and last five rows of the data.

In [18]:
#called the first five rows using .head()
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [19]:
#called the last five rows using .tail()
df.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight Report.Status  \
88884             0.0               NaN                    NaN           NaN   
88885             0.0               NaN                    NaN           NaN   
88886             1.0               VMC                    NaN           NaN   
88887             0.0               NaN                    NaN           NaN   
88888             1.0               NaN                    NaN           NaN   

      Publication.Date  
88884       29-12-2022  
88885              NaN  
88886       27-12-2022  
88887              NaN  
88888       30-12-2022  

[5 rows x 31 columns]

In [20]:
#used the .info() to have an over view of the entire dataset(dtypes,missing values,overall structures)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50130 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

The dataset is a pandas DataFrame that has a combination of objects and floats ; 5 columns are floats and 26 columns are objects.

We also get to see the columns with the most number of missing values and the ones that have less missing values.

In [23]:
#use the .shape to check the shape of the dataset
df.shape

(88889, 31)

The Aviation dataset has 88889 rows and 31 columns, making it a two dimensional matrix

In [25]:
#use the .column fnction to check the columns in the data
df.columns
#This gives us the total columns taht are in the dataset

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [26]:
#do the .describes to get an overview of the data
df.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

Only the numeric columns are affected/inputed using the method. We see the count,mean,minimum and maximum values.

**b) Handling Missing Values and Inconsistent Data**

We start by checkiing the null and placeholder values in the dataset.

In [30]:
#using .isnull() and .sum() to check the null values and sum of the null values
df.isnull()
#returns a bool (True if it's null and False if it's not)

Event.Id  Investigation.Type  Accident.Number  Event.Date  Location  \
0         False               False            False       False     False   
1         False               False            False       False     False   
2         False               False            False       False     False   
3         False               False            False       False     False   
4         False               False            False       False     False   
...         ...                 ...              ...         ...       ...   
88884     False               False            False       False     False   
88885     False               False            False       False     False   
88886     False               False            False       False     False   
88887     False               False            False       False     False   
88888     False               False            False       False     False   

       Country  Latitude  Longitude  Airport.Code  Airport.Name  ...  \
0        False      True       True          True          True  ...   
1        False      True       True          True          True  ...   
2        False     False      False          True          True  ...   
3        False      True       True          True          True  ...   
4        False      True       True          True          True  ...   
...        ...       ...        ...           ...           ...  ...   
88884    False      True       True          True          True  ...   
88885    False      True       True          True          True  ...   
88886    False     False      False         False         False  ...   
88887    False      True       True          True          True  ...   
88888    False      True       True          True          True  ...   

       Purpose.of.flight  Air.carrier  Total.Fatal.Injuries  \
0                  False         True                 False   
1                  False         True                 False   
2                  False         True                 False   
3                  False         True                 False   
4                  False         True                 False   
...                  ...          ...                   ...   
88884              False         True                 False   
88885               True         True                 False   
88886              False         True                 False   
88887              False        False                 False   
88888              False         True                 False   

       Total.Serious.Injuries  Total.Minor.Injuries  Total.Uninjured  \
0                       False                 False            False   
1                       False                 False            False   
2                        True                  True             True   
3                       False                 False            False   
4                       False                  True            False   
...                       ...                   ...              ...   
88884                   False                 False            False   
88885                   False                 False            False   
88886                   False                 False            False   
88887                   False                 False            False   
88888                   False                 False            False   

       Weather.Condition  Broad.phase.of.flight  Report.Status  \
0                  False                  False          False   
1                  False                  False          False   
2                  False                  False          False   
3                  False                  False          False   
4                  False                  False          False   
...                  ...                    ...            ...   
88884               True                   True           True   
88885               True                   True        

In [31]:
#sum the null values
df.isnull().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38759
Airport.Name              36185
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1382
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7096
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6384
Publication.Date          13771
dtype: i

This helps us in knowing which columns are likely to be dropped since they contain so many null values and replacing them will affect our findings or analysis.

In [33]:
#replace the null values in least affected columns
location_mode= df["Location"].mode()[0]
df["Location"]=df["Location"].fillna(location_mode)
df["Location"].head()

0    MOOSE CREEK, ID
1     BRIDGEPORT, CA
2      Saltville, VA
3         EUREKA, CA
4         Canton, OH
Name: Location, dtype: object

Replaced all null values with the mode.

In [35]:
#Do the same for all the others
Make_mode= df["Make"].mode()[0]
df["Make"]=df["Make"].fillna(Make_mode)
df["Make"].head()

0     Stinson
1       Piper
2      Cessna
3    Rockwell
4      Cessna
Name: Make, dtype: object

In [36]:
#Model of the aircraft
Model_mode= df["Model"].mode()[0]
df["Model"]=df["Model"].fillna(Model_mode)
df["Model"].head()

0       108-3
1    PA24-180
2        172M
3         112
4         501
Name: Model, dtype: object

In [37]:
#Injury Severity
Injury_Severity= df["Injury.Severity"].mode()[0]
df["Injury.Severity"]=df["Injury.Severity"].fillna(Injury_Severity)
df["Injury.Severity"].head()

0    Fatal(2)
1    Fatal(4)
2    Fatal(3)
3    Fatal(2)
4    Fatal(1)
Name: Injury.Severity, dtype: object

In [38]:
#For aircraft damage
Aircraft_damage = df["Aircraft.damage"].mode()[0]
df["Aircraft.damage"]=df["Aircraft.damage"].fillna(Aircraft_damage)
df["Aircraft.damage"].head()

0    Destroyed
1    Destroyed
2    Destroyed
3    Destroyed
4    Destroyed
Name: Aircraft.damage, dtype: object

In [39]:
#Change for the amature built as well
Amateur_built= df['Amateur.Built'].mode()[0]
df['Amateur.Built']= df['Amateur.Built'].fillna(Amateur_built)
df['Amateur.Built'].head()

0    No
1    No
2    No
3    No
4    No
Name: Amateur.Built, dtype: object

In [40]:
df['Weather.Condition'].unique()

array(['UNK', 'IMC', 'VMC', nan, 'Unk'], dtype=object)

Weather Condition has null values and placeholder values(UNK,Unk) that need to be replaced. First we're going to convert them to null values
then replace them with the "mode" weather condition as we don't want to drop any data.

In [42]:
#Convert all the strings to uppercase
df["Weather.Condition"] = df["Weather.Condition"].str.upper()

In [43]:
#convert the unknown weather conditions to null 
df["Weather.Condition"].replace("UNK", np.nan, inplace=True)

In [44]:
#The unique weather conditions are now VMC and IMC; with NaN that have been converted.
df["Weather.Condition"].unique()

array([nan, 'IMC', 'VMC'], dtype=object)

In [45]:
#Replace all the nulls with the mode
Weather_mode= df["Weather.Condition"].mode()[0]
df["Weather.Condition"]= df["Weather.Condition"].fillna(Weather_mode)

In [46]:
df['Weather.Condition'].unique()

array(['VMC', 'IMC'], dtype=object)

In [47]:
df['Purpose.of.flight'].unique()

array(['Personal', nan, 'Business', 'Instructional', 'Unknown', 'Ferry',
       'Executive/corporate', 'Aerial Observation', 'Aerial Application',
       'Public Aircraft', 'Skydiving', 'Other Work Use', 'Positioning',
       'Flight Test', 'Air Race/show', 'Air Drop',
       'Public Aircraft - Federal', 'Glider Tow',
       'Public Aircraft - Local', 'External Load',
       'Public Aircraft - State', 'Banner Tow', 'Firefighting',
       'Air Race show', 'PUBS', 'ASHO', 'PUBL'], dtype=object)

In [48]:
df["Purpose.of.flight"].replace("Unknown",np.nan, inplace=True)#replace Unknown with nan
Purpose_flight= df["Purpose.of.flight"].mode()[0]
df["Purpose.of.flight"]= df["Purpose.of.flight"].fillna(Purpose_flight)#replace all nulls with the mode 
df["Purpose.of.flight"].head()

0    Personal
1    Personal
2    Personal
3    Personal
4    Personal
Name: Purpose.of.flight, dtype: object

In [49]:
df["Number.of.Engines"] = pd.to_numeric(df["Number.of.Engines"], errors="coerce")
df["Number.of.Engines"].unique()

array([ 1., nan,  2.,  0.,  3.,  4.,  8.,  6.])

In [50]:
df["Number.of.Engines"].isnull().describe()

count     88889
unique        2
top       False
freq      82805
Name: Number.of.Engines, dtype: object

In [51]:
No_of_engines= df["Number.of.Engines"].mode()[0]
df["Number.of.Engines"] =df["Number.of.Engines"].fillna(No_of_engines)
df["Number.of.Engines"].unique()

array([1., 2., 0., 3., 4., 8., 6.])

In [52]:
df["Engine.Type"].isnull().describe()#get the description of the nulls and non-null values with the frequency of the most present

count     88889
unique        2
top       False
freq      81793
Name: Engine.Type, dtype: object

In [53]:
df["Engine.Type"].unique()#check for unkwons or placeholder values

array(['Reciprocating', nan, 'Turbo Fan', 'Turbo Shaft', 'Unknown',
       'Turbo Prop', 'Turbo Jet', 'Electric', 'Hybrid Rocket',
       'Geared Turbofan', 'LR', 'NONE', 'UNK'], dtype=object)

In [54]:
df["Engine.Type"].replace(["UNK", "NONE"], np.nan, inplace=True)#replace placeholder values with null values
type_engine= df["Engine.Type"].mode()[0]
df["Engine.Type"] =df["Engine.Type"].fillna(type_engine)#replace the nulls with the mode engine type
df["Engine.Type"].unique()

array(['Reciprocating', 'Turbo Fan', 'Turbo Shaft', 'Unknown',
       'Turbo Prop', 'Turbo Jet', 'Electric', 'Hybrid Rocket',
       'Geared Turbofan', 'LR'], dtype=object)

In [55]:
df["Total.Fatal.Injuries"] = pd.to_numeric(df["Total.Fatal.Injuries"], errors="coerce")
df["Total.Fatal.Injuries"].unique()

array([  2.,   4.,   3.,   1.,  nan,   0.,   8.,  78.,   7.,   6.,   5.,
       153.,  12.,  14.,  23.,  10.,  11.,   9.,  17.,  13.,  29.,  70.,
       135.,  31., 256.,  25.,  82., 156.,  28.,  18.,  43.,  15., 270.,
       144., 174., 111., 131.,  20.,  73.,  27.,  34.,  87.,  30.,  16.,
        47.,  56.,  37., 132.,  68.,  54.,  52.,  65.,  72., 160., 189.,
       123.,  33., 110., 230.,  97., 349., 125.,  35., 228.,  75., 104.,
       229.,  80., 217., 169.,  88.,  19.,  60., 113., 143.,  83.,  24.,
        44.,  64.,  92., 118., 265.,  26., 138., 206.,  71.,  21.,  46.,
       102., 115., 141.,  55., 121.,  45., 145., 117., 107., 124.,  49.,
       154.,  96., 114., 199.,  89.,  57., 152.,  90., 103., 158., 157.,
        42.,  77., 127.,  50., 239., 295.,  58., 162., 150., 224.,  62.,
        66., 112., 188.,  41., 176.])

In [56]:
#replacing all the nulls with the mode using the lambda function
df["Total.Fatal.Injuries"] = df["Total.Fatal.Injuries"].apply(lambda x: df["Total.Fatal.Injuries"].mode()[0] if pd.isna(x) else x)
df["Total.Fatal.Injuries"].unique()

array([  2.,   4.,   3.,   1.,   0.,   8.,  78.,   7.,   6.,   5., 153.,
        12.,  14.,  23.,  10.,  11.,   9.,  17.,  13.,  29.,  70., 135.,
        31., 256.,  25.,  82., 156.,  28.,  18.,  43.,  15., 270., 144.,
       174., 111., 131.,  20.,  73.,  27.,  34.,  87.,  30.,  16.,  47.,
        56.,  37., 132.,  68.,  54.,  52.,  65.,  72., 160., 189., 123.,
        33., 110., 230.,  97., 349., 125.,  35., 228.,  75., 104., 229.,
        80., 217., 169.,  88.,  19.,  60., 113., 143.,  83.,  24.,  44.,
        64.,  92., 118., 265.,  26., 138., 206.,  71.,  21.,  46., 102.,
       115., 141.,  55., 121.,  45., 145., 117., 107., 124.,  49., 154.,
        96., 114., 199.,  89.,  57., 152.,  90., 103., 158., 157.,  42.,
        77., 127.,  50., 239., 295.,  58., 162., 150., 224.,  62.,  66.,
       112., 188.,  41., 176.])

In [57]:
df["Total.Serious.Injuries"] = pd.to_numeric(df["Total.Serious.Injuries"], errors="coerce")#convert no numbers
df["Total.Serious.Injuries"].unique()

array([  0.,  nan,   2.,   1.,   6.,   4.,   5.,  10.,   3.,   8.,   9.,
         7.,  15.,  17.,  28.,  26.,  47.,  14.,  81.,  13., 106.,  60.,
        16.,  21.,  50.,  44.,  18.,  12.,  45.,  39.,  43.,  11.,  25.,
        59.,  23.,  55.,  63.,  88.,  41.,  34.,  53.,  33.,  67.,  35.,
        20., 137.,  19.,  27., 125., 161.,  22.])

In [58]:
#replacing all the nulls with the mode using the lambda fnct
df["Total.Serious.Injuries"] = df["Total.Serious.Injuries"].apply(lambda x: df["Total.Serious.Injuries"].mode()[0] if pd.isna(x) else x)
df["Total.Serious.Injuries"].unique()

array([  0.,   2.,   1.,   6.,   4.,   5.,  10.,   3.,   8.,   9.,   7.,
        15.,  17.,  28.,  26.,  47.,  14.,  81.,  13., 106.,  60.,  16.,
        21.,  50.,  44.,  18.,  12.,  45.,  39.,  43.,  11.,  25.,  59.,
        23.,  55.,  63.,  88.,  41.,  34.,  53.,  33.,  67.,  35.,  20.,
       137.,  19.,  27., 125., 161.,  22.])

In [59]:
df["Total.Minor.Injuries"] = pd.to_numeric(df["Total.Minor.Injuries"], errors="coerce")
df["Total.Minor.Injuries"].unique()

array([  0.,  nan,   1.,   3.,   2.,   4.,  24.,   6.,   5.,  25.,  17.,
        19.,  33.,  14.,   8.,  13.,  15.,   7.,   9.,  16.,  20.,  11.,
        12.,  10.,  38.,  42.,  29.,  62.,  28.,  31.,  39.,  32.,  18.,
        27.,  57.,  50.,  23., 125.,  45.,  26.,  36.,  69.,  21.,  96.,
        30.,  22.,  58., 171.,  65.,  71., 200.,  68.,  47., 380.,  35.,
        43.,  84.,  40.])

In [60]:
#replacing all the nulls with the mode using the lambda function
df["Total.Minor.Injuries"] = df["Total.Minor.Injuries"].apply(lambda x: df["Total.Minor.Injuries"].mode()[0] if pd.isna(x) else x)
df["Total.Minor.Injuries"].unique()

array([  0.,   1.,   3.,   2.,   4.,  24.,   6.,   5.,  25.,  17.,  19.,
        33.,  14.,   8.,  13.,  15.,   7.,   9.,  16.,  20.,  11.,  12.,
        10.,  38.,  42.,  29.,  62.,  28.,  31.,  39.,  32.,  18.,  27.,
        57.,  50.,  23., 125.,  45.,  26.,  36.,  69.,  21.,  96.,  30.,
        22.,  58., 171.,  65.,  71., 200.,  68.,  47., 380.,  35.,  43.,
        84.,  40.])

In [61]:
df["Total.Uninjured"] = pd.to_numeric(df["Total.Uninjured"], errors="coerce")#to numeric value
df["Total.Uninjured"].unique()

array([  0.,  nan,  44.,   2.,   1.,   3.,   6.,   4., 149.,  12., 182.,
       154.,   5.,  10.,   7., 119.,  36.,  51.,  16.,  83.,   9.,  68.,
        30.,  20.,  18.,   8., 108.,  11., 152.,  21.,  48.,  56., 113.,
       129., 109.,  29.,  13.,  84.,  74., 142., 102., 393., 128., 112.,
        17.,  65.,  67., 136.,  23., 116.,  22.,  57.,  58.,  73., 203.,
        31., 201., 412., 159.,  39., 186., 588.,  82.,  95., 146., 190.,
       245., 172.,  52.,  25.,  59., 131., 151., 180., 150.,  86.,  19.,
       133., 240.,  15., 145., 125., 440.,  77., 122., 205., 289., 110.,
        79.,  66.,  87.,  78.,  49., 104., 250.,  33., 138., 100.,  53.,
       158., 127., 160., 260.,  47.,  38., 165., 495.,  81.,  41.,  14.,
        72.,  98., 263., 188., 239.,  27., 105., 111., 212., 157.,  46.,
       121.,  75.,  71.,  45.,  91.,  99.,  85.,  96.,  50.,  93., 276.,
       365., 371., 200., 103., 189.,  37., 107.,  61.,  26., 271., 130.,
        89., 439., 132., 219.,  43., 238., 195., 11

In [62]:
#replacing all the nulls with the mode using the lambda fnct
df["Total.Uninjured"] = df["Total.Uninjured"].apply(lambda x: df["Total.Uninjured"].mode()[0] if pd.isna(x) else x)
df["Total.Uninjured"].unique()

array([  0.,  44.,   2.,   1.,   3.,   6.,   4., 149.,  12., 182., 154.,
         5.,  10.,   7., 119.,  36.,  51.,  16.,  83.,   9.,  68.,  30.,
        20.,  18.,   8., 108.,  11., 152.,  21.,  48.,  56., 113., 129.,
       109.,  29.,  13.,  84.,  74., 142., 102., 393., 128., 112.,  17.,
        65.,  67., 136.,  23., 116.,  22.,  57.,  58.,  73., 203.,  31.,
       201., 412., 159.,  39., 186., 588.,  82.,  95., 146., 190., 245.,
       172.,  52.,  25.,  59., 131., 151., 180., 150.,  86.,  19., 133.,
       240.,  15., 145., 125., 440.,  77., 122., 205., 289., 110.,  79.,
        66.,  87.,  78.,  49., 104., 250.,  33., 138., 100.,  53., 158.,
       127., 160., 260.,  47.,  38., 165., 495.,  81.,  41.,  14.,  72.,
        98., 263., 188., 239.,  27., 105., 111., 212., 157.,  46., 121.,
        75.,  71.,  45.,  91.,  99.,  85.,  96.,  50.,  93., 276., 365.,
       371., 200., 103., 189.,  37., 107.,  61.,  26., 271., 130.,  89.,
       439., 132., 219.,  43., 238., 195., 118., 17

In [63]:
df["Registration.Number"].isna().describe()#87507 non-null values

count     88889
unique        2
top       False
freq      87507
Name: Registration.Number, dtype: object

In [64]:
#used the .apply() method with the lambda function
df["Registration.Number"]= df["Registration.Number"].apply(lambda x: df["Registration.Number"].mode()[0] if pd.isna(x) else x)
df["Registration.Number"].head()

0    NC6404
1    N5069P
2    N5142R
3    N1168J
4     N15NY
Name: Registration.Number, dtype: object

In [65]:
df["Broad.phase.of.flight"].unique()

array(['Cruise', 'Unknown', 'Approach', 'Climb', 'Takeoff', 'Landing',
       'Taxi', 'Descent', 'Maneuvering', 'Standing', 'Go-around', 'Other',
       nan], dtype=object)

In [66]:
df["Broad.phase.of.flight"].replace(["Unknown","Other"],np.nan, inplace=True)#converting placehoolder values to null values for easier replacement
#replacing all the nulls with the mode using the lambda function
df["Broad.phase.of.flight"] = df["Broad.phase.of.flight"].apply(lambda x: df["Broad.phase.of.flight"].mode()[0] if pd.isna(x) else x)
df["Broad.phase.of.flight"].unique()

array(['Cruise', 'Landing', 'Approach', 'Climb', 'Takeoff', 'Taxi',
       'Descent', 'Maneuvering', 'Standing', 'Go-around'], dtype=object)

In [67]:
df["Airport.Code"].unique()

array([nan, 'N58', 'JAX', ..., 'SKMD', 'OMAA', 'EIKH'], dtype=object)

In [68]:
df["Airport.Code"].replace([np.nan, "NONE"], "Missing", inplace=True)#replace the nulls and NONE with a string "missing"
df["Airport.Code"].value_counts().head()

Airport.Code
Missing    40247
PVT          485
APA          160
ORD          149
MRI          137
Name: count, dtype: int64

In [69]:
df["Airport.Name"].value_counts()

Airport.Name
Private                           240
PRIVATE                           224
Private Airstrip                  153
NONE                              146
PRIVATE STRIP                     111
                                 ... 
STEHEKIN                            1
OTTAWA MUNICIPAL                    1
DWIGHT AIRPORT                      1
PHOENIX DEER VALLEY MUNI            1
WICHITA DWIGHT D EISENHOWER NT      1
Name: count, Length: 24870, dtype: int64

In [70]:
df["Airport.Name"].replace([np.nan, "NONE"], "Missing", inplace=True)#replace the name with missing cause the numbers are too
#high to change(36331).If we replace they're going to have a huge impact on the data, so I choose to replace with a string.
df["Airport.Name"].value_counts().head()

Airport.Name
Missing             36331
Private               240
PRIVATE               224
Private Airstrip      153
PRIVATE STRIP         111
Name: count, dtype: int64

In [71]:
df["Publication.Date"].unique()

array([nan, '19-09-1996', '26-02-2007', ..., '22-12-2022', '23-12-2022',
       '29-12-2022'], dtype=object)

In [72]:
df["Publication.Date"].replace([np.nan], "Missing", inplace=True)#replace the nulls with "missing".
df["Publication.Date"].unique()

array(['Missing', '19-09-1996', '26-02-2007', ..., '22-12-2022',
       '23-12-2022', '29-12-2022'], dtype=object)

In [140]:
df["Country"] = df["Country"].apply(lambda x: df["Country"].mode()[0] if pd.isna(x) else x)
df["Country"].unique()

array(['United States', 'GULF OF MEXICO', 'Puerto Rico', 'ATLANTIC OCEAN',
       'HIGH ISLAND', 'Bahamas', 'MISSING', 'Pakistan', 'Angola',
       'Germany', 'Korea, Republic Of', 'Martinique', 'American Samoa',
       'PACIFIC OCEAN', 'Canada', 'Bolivia', 'Mexico', 'Dominica',
       'Netherlands Antilles', 'Iceland', 'Greece', 'Guam', 'Australia',
       'CARIBBEAN SEA', 'West Indies', 'Japan', 'Philippines',
       'Venezuela', 'Bermuda', 'San Juan Islands', 'Colombia',
       'El Salvador', 'United Kingdom', 'British Virgin Islands',
       'Netherlands', 'Costa Rica', 'Mozambique', 'Jamaica', 'Panama',
       'Guyana', 'Norway', 'Hong Kong', 'Portugal', 'Malaysia',
       'Turks And Caicos Islands', 'Northern Mariana Islands',
       'Dominican Republic', 'Suriname', 'Honduras', 'Congo', 'Belize',
       'Guatemala', 'Anguilla', 'France', 'St Vincent And The Grenadines',
       'Haiti', 'Montserrat', 'Papua New Guinea', 'Cayman Islands',
       'Sweden', 'Taiwan', 'Senegal', 'Bar

In [73]:
#to convert all the dtypes to the appropriate dtypes automatically
df.convert_dtypes()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude  Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN        NaN      Missing   
1       BRIDGEPORT, CA  United States        NaN        NaN      Missing   
2        Saltville, VA  United States  36.922223 -81.878056      Missing   
3           EUREKA, CA  United States        NaN        NaN      Missing   
4           Canton, OH  United States        NaN        NaN      Missing   
...                ...            ...        ...        ...          ...   
88884    Annapolis, MD  United States        NaN        NaN      Missing   
88885      Hampton, NH  United States        NaN        NaN      Missing   
88886       Payson, AZ  United States    341525N   1112021W          PAN   
88887       Morgan, UT  United States        NaN        NaN      Missing   
88888       Athens, GA  United States        NaN        NaN      Missing   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0          Missing  ...          Personal                <NA>   
1          Missing  ...          Personal                <NA>   
2          Missing  ...          Personal                <NA>   
3          Missing  ...          Personal                <NA>   
4          Missing  ...          Personal                <NA>   
...            ...  ...               ...                 ...   
88884      Missing  ...          Personal                <NA>   
88885      Missing  ...          Personal                <NA>   
88886       PAYSON  ...          Personal                <NA>   
88887      Missing  ...          Personal  MC CESSNA 210N LLC   
88888      Missing  ...          Personal                <NA>   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                        2                      0                    0   
1                        4                      0                    0   
2                        3                      0                    0   
3                        2                      0                    0   
4                        1                      2                    0   
...                    ...                    ...                  ...   
88884                    0                      1                    0   
88885                    0                      0                    0   
88886                    0                      0                    0   
88887                    0                      0                    0   
88888                    0                      1                    0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                   0               VMC                 Cruise   
1                   0               VMC                Landing   
2                   0               IMC                 Cruise   
3                   0               IMC                 Cruise   
4                   0               VMC               Approach   
...               ...               ...                    ...   
88884               0               VMC                Landing   
88885               0               V

In [74]:
print(df.dtypes)

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88889 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            88889 non-null  object 
 9   Airport.Name            88889 non-null  object 
 10  Injury.Severity         88889 non-null  object 
 11  Aircraft.damage         88889 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     88889 non-null  object 
 14  Make                    88889 non-null

In [138]:
#convert date to real time
df["Publication.Date"] = pd.to_datetime(df["Publication.Date"],format="%Y-%m-%d", errors= "coerce")
df["Event.Date"] = pd.to_datetime(df["Event.Date"], errors= "coerce")
df["Event.Date"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 88889 entries, 0 to 88888
Series name: Event.Date
Non-Null Count  Dtype         
--------------  -----         
88889 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 694.6 KB


In [134]:
df['Latitude'] = df['Latitude'].replace(np.nan,'missing data')
df['Longitude'] = df['Longitude'].replace(np.nan,'missing data')
df.head()

Event.Id Investigation.Type Accident.Number Event.Date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          Location        Country      Latitude     Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States  missing data  missing data      Missing   
1   BRIDGEPORT, CA  United States  missing data  missing data      Missing   
2    Saltville, VA  United States     36.922223    -81.878056      Missing   
3       EUREKA, CA  United States  missing data  missing data      Missing   
4       Canton, OH  United States  missing data  missing data      Missing   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0      Missing  ...          Personal         NaN                  2.0   
1      Missing  ...          Personal         NaN                  4.0   
2      Missing  ...          Personal         NaN                  3.0   
3      Missing  ...          Personal         NaN                  2.0   
4      Missing  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    0.0                  0.0             0.0   
3                    0.0                  0.0             0.0   
4                    2.0                  0.0             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               VMC                 Cruise  Probable Cause          Missing  
1               VMC                Landing  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]